### PR TITLE
Implement SSO for Plesk

### PR DESCRIPTION
**Pull request by azizozbek**

### CODE REVIEW

File: src/modules/Servicehosting/Api/Client.php [hunk 1]
```diff
@@ -51,6 +51,20 @@ public function change_password($data)
...
     }
 
+    /**
+     * Create plesk session.
+     *
+     * @param int    $order_id  - Hosting account order id
+     *
+     * @return bool
+     */
+    public function plesk_login($data)
```
Comment (jaapmarcus) [inline src/modules/Servicehosting/Api/Client.php:61]: We should "Generalised" the command so they are reusable for other panels...

Comment (jaapmarcus) [reply src/modules/Servicehosting/Api/Client.php:61]: And for the other locations where Plesk is used ...

**Review by BelleNottelling — CHANGES_REQUESTED** [review]
Hi there, thanks for the PR!

It's appreciated, however we cannot accept this in it's current form since the implementation is built to be specifically for Plesk. We need a implementation that's generic and that can be implemented by any of the given server managers.

One possible option is expanding the `getLoginUrl` function to receive the Server_Account instance & then build the link utilizing that to make the login link. That way a server manager can either continue to provide the login link or if it supports such functionality, it can provide an SSO link

**Comment by azizozbek** [comment]
thnx for the reviews, i will look at it how we could generalise the flow. currently i have only plesk to test, if anyone knows the api docs of other server management tools or test account, it would be great to know

**Comment by jaapmarcus** [comment]
> thnx for the reviews, i will look at it how we could generalise the flow. currently i have only plesk to test, if anyone knows the api docs of other server management tools or test account, it would be great to know

We need to rename things so it could be used generally and update the server adapter it self to support sso. So in the future other users can implement their own system....

**Comment by azizozbek** [comment]
any review?

**Comment by admdly** [comment]
> any review?

I see @BelleNottelling has already engaged with you, so I will leave it for them to further review. 

However, note that this is not currently passing the CI checks - these should all be passing before requesting review.

**Comment by azizozbek** [comment]
I had problems also when i was installing the fossbilling manually:
https://fossbilling.org/docs/getting-started/building
I didn't make any change in composer.json / lock. So it would be great if someone give a look, thx.

**Comment by BelleNottelling** [comment]
I rarely review things that still have failing checks as that to mean that the PR is likely still incomplete.

> I had problems with "composer install" also when i was installing the fossbilling manually

I'm sorry, but we can't really help you unless you give details on what problems you had. 
The preview workflow is building correctly so the best I can say is that the issue seems to be specific to your setup

**Comment by admdly** [comment]
> I had problems with "composer install" also when i was installing the fossbilling manually: https://fossbilling.org/docs/getting-started/building I didn't make any change in composer.json / lock. So it would be great if someone give a look, thx.

Could you advise what the issues/errors are? It's difficult to comment without knowing what is going wrong.

**Comment by azizozbek** [comment]
sorry i totally forgot about tests, now it should be solved :)

**Comment by BelleNottelling** [comment]
Okay, I've given a good look over the changes and my initial thought is that this can be done cleaner while retaining the current implementation, however there are two variants of implementations that I can think of:

1. We implement a "login" API endpoint that accepts an order ID. That will fetch the correct info from the DB and then pass that along to a server manager's `getLoginUrl` function. If the server manager supports SSO, it'll use the info provided to create an SSO link and return it, but if it doesn't then it just returns a link to the login page as they currently do now. The login endpoint will then take that link and send it to the client as a temporary redirect header (to prevent it from being cached by the browser). This method would reduce the number of similar functions that need to be implemented and removes to the need to ask a server manager if it supports SSO. The downside is that this would be another breaking change and would either need to happen basically at this very moment or it will have to wait until 0.7.0.
2. The implementation you've done here were server managers need to define 3 different functions and it also needs to be checked within the panel. This is functionally fine, I just consider to to be less optimal implementation. Ultimately if other team members are OK with that choice then I am also OK with it, but only once you clean it up more as things can be done so that server managers which lack SSO support don't need to define `getLoginUrl`, `generateSsoLink`, and `ssoSupport` all at the same time.

Personally, I would much rather do option number 1 and if everyone else agrees I am willing to spin up a new PR that implements that and your [SSO login code for the Plesk adapter](https://github.com/FOSSBilling/FOSSBilling/pull/1842/files#diff-75f0771823f6c6d83faea7fc79b4925b6e0b6f43a481452c6fdb2f1f9394fd64R49-R52) with you as a co-author and then get that to be one of the final items for 0.6.0.

Otherwise, I will probably say that this needs to wait and be a part of 0.6.1 as we are basically in the final week of changes and at this stage we really only want to be merging either bug-fixes or some of the very final items for the [0.6.0 release](https://github.com/orgs/FOSSBilling/projects/2) (which even then, some of those are somewhat likely to get delayed until a later release).

@admdly, @John-S4, @jaapmarcus, and @evrifaessa please share your thoughts

**Comment by azizozbek** [comment]
Thnx for review. 
I think the "getLoginUrl" should only return the ServerUrl to output or redirect. One function should have only one task, so it's independent executable/callable. 
Either way we have to check if the ServerManager supports SSO, that's why i don't see a redundant here with 3 Functions.
getLoginUrl() => the server url
ssoSupport => checks if the server manager supports sso. could be stored in db to detach this function.
generateSsoLink => generates a link for 1minute.

**Comment by BelleNottelling** [comment]
> I think the "getLoginUrl" should only return the ServerUrl to output or redirect. One function should have only one task, so it's independent executable/callable.

With what I suggested, it does exactly that. It turns a link and the *system* handles the redirect when needed, not the function. But at the end of the day `getLoginUrl` and `generateSsoLink` in this PR are two seperate functions which are handling virtually identical purposes. Both give a link that will get the user logged into the control panel. One option simply bypasses the need for them to enter their credentials manually. 

Separately as you have proposed means the system needs to have separate code paths and behavior that allows it to do nearly the same thing in two different ways which is simply unneeded complexity. If we pass the order info to `getLoginUrl` a sever manager that supports SSO can generate an SSO link and return that. If it doesn't support SSO then it has to do nothing other than return a link to the login page much like they do now. Merging them *reduces* branching behavior and complexity for *both* the system and the server manger.

>  Either way we have to check if the ServerManager supports SSO, that's why i don't see a redundant here with 3 Functions.

I am sorry, but this is simply wrong. If the `Server_Manager` class is setup correctly the `ssoSupport` and `generateSsoLink` functions can have default returns for when a server does not support SSO link generation. Then the only time a server manager would need to define all 3 is it does support SSO link generation. The other server managers would not need to do so & this would reduce what is otherwise completely unnecessary clutter in those server managers.

**Comment by azizozbek** [comment]
in your suggestion the usecase would be;
- "Login" button redirects to intern api andpoint/controller 
- This endpoint calls "getLoginUrl" 
- getLoginUrl redirects either with SSO or without
So the ssoSupport and generateSsoLink will be detached. Correct?

**Comment by BelleNottelling** [comment]
Kind of, except with what I proposed we shouldn't need `ssoSupport` or `generateSsoLink`.
At the absolute most if we need a way to generate non SSO links we can either add an optional parameter to the function or build it to expect `null` in place of the `Server_Account` parameter. Either way gives us an expected behavior to generate a non-SSO link.

But you don't need to go and change your PR right now.
I intentionally opened the conversation up to other team members. If they are happy with your solution, all I want is for the little bit of cleanup that's needed to remove the `ssoSupport` and `generateSsoLink` functions from server managers that don't support SSO link generation.

If the other team members prefer what I proposed. I can also handle that work and make a PR that does that + the code you've created to get the plesk SSO going, so that way you still don't need to completely redo your PR again.

**Comment by John-S4** [comment]
> 1. We implement a "login" API endpoint that accepts an order ID...

@BelleNottelling This seems like a sensible option to me, clean and simple. I'd vote for option 1 over option 2.

**Comment by admdly** [comment]
@BelleNottelling has made some great suggestions here, and like @John-S4 I would vote for option 1 over option 2 (assuming the `login` endpoint is authenticated - which I'm sure doesn't need to be said, but I will state anyway for clarity).

@azizozbek I understand the argument about separation of concerns with individual functions, however as with everything there is a balance to be had between simplicity and sticking to a rigid definition of a "function". At the end of the day, there's no reason to add additional complexity simply to stick to some stated programming ideal.

**Comment by azizozbek** [comment]
thnx for reviews. @BelleNottelling would you refactor the code and do the PR or should i do it?

**Comment by BelleNottelling** [comment]
> thnx for reviews. @BelleNottelling would you refactor the code and do the PR or should i do it?

Hi thanks! I'll take care of it and get a PR up in a bit

**Pull request closed without merging**